### PR TITLE
Add github checks

### PR DESCRIPTION
**Pull request by sagudev**
fixes https://github.com/bencherdev/bencher/issues/490

### CODE REVIEW

**Review by epompeii — CHANGES_REQUESTED** [review]
Looking good, @sagudev. Just a couple small comments.

Have you been able to test your changes anywhere?

**Comment by sagudev** [comment]
> Have you been able to test your changes anywhere?

Not yet, hence the draft state.

**Comment by wusyong** [comment]
I tested on a forked test [repo](https://github.com/wusyong/bencher-test) of @dklassic and it can get the [dummy report](https://bencher.dev/console/projects/wusyong-s-project/perf?report=ad9e61f9-5fd3-424f-9140-2af17cd82265&branches=10772c66-8976-475d-aa0e-55c51dfeea41&heads=540f1a33-901b-4eec-83c3-b67fc0d033f9&testbeds=9c2fbc11-08a8-471f-87b3-7146fa3eaf55&benchmarks=a2437b12-42a7-43a1-ad90-850145ff4081&measures=68c1ecb6-8b6a-4c5d-b657-7acd81d208de&start_time=1727080193000&end_time=1729672193000&key=true&reports_per_page=4&branches_per_page=8&testbeds_per_page=8&benchmarks_per_page=8&plots_per_page=8&reports_page=1&branches_page=1&testbeds_page=1&benchmarks_page=1&plots_page=1).
I think this draft can move into a PR and try to merge it?

**Comment by dklassic** [comment]
So I hope to help pushing this PR forward so I did some testing with the help of @wusyong and @pewsheen with [result here](https://github.com/dklassic/bencher-test/actions/runs/11476462996/job/31936522315), the GitHub Action can now run outside of `pull_request`, `pull_request_target`, and `workflow_run`.

I'm actually not particularly sure if this is sufficient to show that this PR works?

**Comment by sagudev** [comment]
I do not see any github check, so maybe it's not working?

**Comment by dklassic** [comment]
> I do not see any github check, so maybe it's not working?

Just to be sure my understanding is correct, my testing method is basically I made an empty repo with a modified version of [this workflow](https://github.com/servo/servo/blob/875e4dcb652a0f1fcea3ca8a8919cc094d0e0e4b/.github/workflows/linux-bencher.yml) and I run a random push and see if it runs properly. Is this what I'm supposed to be doing?

**Comment by epompeii** [comment]
@dklassic I appreciate the interest in moving this forward.

I would recommend just creating the most basic GitHub Actions workflow possible, a slimmed down version of [the Base branch example](https://bencher.dev/docs/how-to/github-actions/#base-branch).

```yml
on:
  push:
    branches: main

jobs:
  benchmark_base_branch:
    name: Continuous Benchmarking with Bencher
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: sagudev/bencher@github-checks
      - name: Track base branch benchmarks with Bencher
        run: |
          bencher run \
          --project save-walter-white-1234abcd \
          --token '${{ secrets.BENCHER_API_TOKEN }}' \
          --branch main \
          --testbed ubuntu-latest \
          --adapter json \
          --github-actions '${{ secrets.GITHUB_TOKEN }}' \
          bencher mock
 ```
 
 1. Add this workflow file to the `main` branch of your test repo: https://github.com/dklassic/bencher-test
 2. Create a follow up test commit and push it to `dklassic/bencher-test`
 3. There should then be a GitHub Check comment on that commit
 4. Post a link to that commit with the GitHub Check here
 5. If everything works as expected I or @sagudev can pull in your changes to this PR
 6. Party! 🎉

**Comment by dklassic** [comment]
So [a quick test](https://github.com/dklassic/bencher-test/actions/runs/11482013791/job/31954027126) with that slimmed down version of [the Base branch example](https://bencher.dev/docs/how-to/github-actions/#base-branch) leads to the following error:

> Failed to check API version: Failed to get API server version: Failed to send after 10 attempts

I'm not exactly sure what's at fault here and how to resolve this.

**Comment by epompeii** [comment]
> I'm not exactly sure what's at fault here and how to resolve this.

@dklassic this is my mistake. I had `--debug` in my example which will look for a local instance of Bencher instead of Bencher Cloud (`https://api.bencher.dev`). The above example has been corrected and should hopefully work now.

**Comment by dklassic** [comment]
> @dklassic this is my mistake. I had `--debug` in my example which will look for a local instance of Bencher instead of Bencher Cloud (`https://api.bencher.dev`). The above example has been corrected and should hopefully work now.

So a rerun [reports it failed to create a GitHub check](https://github.com/dklassic/bencher-test/actions/runs/11489828032/job/31979351072#step:3:112), I'll look into how to fix it.

**Comment by sagudev** [comment]
> @wusyong noticed that the only thing missing seems to be the [write permission to the workflow](https://github.com/dklassic/bencher-test/commit/8ba27c45d0a367b086740221b0c4e1afaedda0aa) and all is well, I think this PR is now mergable with this simple change.
> 
> Also [working example here](https://github.com/dklassic/bencher-test/actions/runs/11495868828/job/31996525798)

So there are no changes needed apart from rebase? Actually I think it would be better to simply open new PR from your fork (in case I do rebasing wrong).

**Comment by dklassic** [comment]
> So there are no changes needed apart from rebase? Actually I think it would be better to simply open new PR from your fork (in case I do rebasing wrong).

Yeah it seems working as intended and should be mergable. Rebase should be problem free as my repo is just rebased to main branch with Fork then do forced push.

**Comment by sagudev** [comment]
Thanks for testing @dklassic.

**Comment by epompeii** [comment]
Great collaboration!
@wusyong nice catch on the GitHub Actions token permission. Thank you!

I changed the PR [to target `devel`](https://github.com/bencherdev/bencher?tab=readme-ov-file#-all-pull-requests-should-target-the-devel-branch). 

Seeing the final results, I do have two small bits that I would like to have fixed before merging.

1. The name for the job that is created is currently "bencher". I would prefer for this to say "Bencher Report".
2. The part that currently says "Bencher Report" is a bit redundant, as that is repeated by the content itself, so please remove the _current_ part that says "Bencher Report".

**Comment by sagudev** [comment]
> The part that currently says "Bencher Report" is a bit redundant, as that is repeated by the content itself, so please remove the current part that says "Bencher Report".

This will be problematic as API requires that we provide title, I did https://github.com/bencherdev/bencher/pull/511/commits/06bddd6d7b373aacbb06736c63cf581328892d31 as workaround, @dklassic plz retest to make sure everything works as expected.

**Comment by dklassic** [comment]
> This will be problematic as API requires that we provide title, I did [06bddd6](https://github.com/bencherdev/bencher/commit/06bddd6d7b373aacbb06736c63cf581328892d31) as workaround, @dklassic plz retest to make sure everything works as expected.

截圖 2024-10-24 下午6 53 36

[Quick test](https://github.com/dklassic/bencher-test/actions/runs/11497758338/job/32002310584) shows the format breaks hummmmmm

**Comment by dklassic** [comment]
Okay [it's working now](https://github.com/dklassic/bencher-test/actions/runs/11497758338/job/32002727418)!

截圖 2024-10-24 晚上7 02 43

**Comment by epompeii** [comment]
It is important that the "Bencher Report" title is a link to the Report page. Further, I would much prefer the actual Bencher logo to the emoji.

Is it possible to just put an empty string in as the title?

File: services/cli/src/bencher/sub/project/run/ci/github_actions.rs [hunk 1]
```diff
@@ -244,6 +253,43 @@ impl GitHubActions {
...
+        &self,
+        report_comment: &ReportComment,
+        event_str: &str,
+        event: &serde_json::Value,
+    ) -> Result<CheckRun, GitHubError> {
+        let full_name = repository_full_name(event_str, event)?;
+        let (owner, repo) = split_full_name(full_name)?;
+
+        let report = CheckRunOutput {
+            title: "🐰 Bencher Report".to_owned(),
```
Comment (epompeii) [inline services/cli/src/bencher/sub/project/run/ci/github_actions.rs:267]: This needs to be a hyperlink to the Report. So maybe try reverting this and just putting an empty string here?

**Review by epompeii — CHANGES_REQUESTED** [review]

File: services/cli/src/bencher/sub/project/run/ci/github_actions.rs (hunk 1 above)
Comment (sagudev) [reply services/cli/src/bencher/sub/project/run/ci/github_actions.rs]: Let's try empty string (ping @dklassic), but if it does not work, this is the best we can get (title is escaped so no md/html is possible).

Comment (dklassic) [reply services/cli/src/bencher/sub/project/run/ci/github_actions.rs]: [Tested and yeah it works](https://github.com/dklassic/bencher-test/actions/runs/11497758338/job/32008653984).

Comment (sagudev) [reply services/cli/src/bencher/sub/project/run/ci/github_actions.rs]: Empty title still causes empty `<h2>` to be created, which causes another _______ line:
slika

@epompeii what do you prefer?

Comment (epompeii) [reply services/cli/src/bencher/sub/project/run/ci/github_actions.rs]: @sagudev I would prefer the extra, useless `<h2>` and it's ensuing extra `<hr />`.

So yes, please revert that commit and let's just move over to using the empty string.

Comment (sagudev) [reply services/cli/src/bencher/sub/project/run/ci/github_actions.rs]: Sorry for late response, I got this notification in my spam folder 😞.

File: services/cli/src/bencher/sub/project/run/ci/github_actions.rs [hunk 2]
```diff
@@ -187,11 +195,12 @@ impl GitHubActions {
                 self.ci_number.ok_or(GitHubError::NoWorkflowRunPRNumber)?
             },
             _ => {
-                cli_println_quietable!(
-                    log,
-                    "Not running as an expected GitHub Action event (`pull_request`, `pull_request_target`, or `workflow_run`). Skipping CI integration.\n{}",
+                cli_println!(
```
Comment (epompeii) [inline services/cli/src/bencher/sub/project/run/ci/github_actions.rs:198]: This needs to still be `cli_println_quietable`. See: https://bencher.dev/docs/explanation/bencher-run/#--quiet

**Review by epompeii — CHANGES_REQUESTED** [review]

**Review by epompeii — APPROVED** [review]
Looks great! Thank you, @sagudev.

**Comment by epompeii** [comment]
An additional thank you to @dklassic and @wusyong for your help testing things out.
I have already update the docs to include the new required permission: https://bencher.dev/docs/how-to/github-actions/#base-branch

**Pull request merged**